src/stores/WatchHistoryController.ts

```diff
@@ -51,38 +48,52 @@ export const persistWatchHistory = () => {
   const { user } = useAccountStore.getState();
 
   if (user) {
-    return updatePersonalShelves();
+    updatePersonalShelves();
   }
 
-  return persist.setItem(PERSIST_KEY_WATCH_HISTORY, serializeWatchHistory(watchHistory));
+  persist.setItem(PERSIST_KEY_WATCH_HISTORY, serializeWatchHistory(watchHistory));
 };
 
-export const createWatchHistoryItem = (item: PlaylistItem, videoProgress: VideoProgress): WatchHistoryItem => {
+export const createWatchHistoryItem = (item: PlaylistItem, videoProgress: number): WatchHistoryItem => {
   return {
     mediaid: item.mediaid,
     title: item.title,
     tags: item.tags,
-    duration: videoProgress.duration,
-    progress: videoProgress.progress,
+    duration: item.duration,
+    progress: videoProgress,
     playlistItem: item,
   } as WatchHistoryItem;
 };
 
-export const saveItem = (item: PlaylistItem, videoProgress: VideoProgress | null) => {
+/**
+ *  If we already have an element with continue watching state, we:
+ *    1. Update the progress
+ *    2. Move the element to the continue watching list start
+ *  Otherwise:
+ *    1. Move the element to the continue watching list start
+ *    2. If there are many elements in continue watching state we remove the oldest one
+ */
+export const saveItem = (item: PlaylistItem, videoProgress: number | null) => {
   const { watchHistory } = useWatchHistoryStore.getState();
 
   if (!videoProgress) return;
 
   const watchHistoryItem = createWatchHistoryItem(item, videoProgress);
   const index = watchHistory.findIndex(({ mediaid }) => mediaid === watchHistoryItem.mediaid);
 
+  let updatedHistory = watchHistory;
+
   if (index > -1) {
-    watchHistory[index] = watchHistoryItem;
-    useWatchHistoryStore.setState({ watchHistory });
+    updatedHistory = [watchHistoryItem, ...watchHistory.filter(({ mediaid }) => mediaid !== watchHistoryItem.mediaid)];
   } else {
-    useWatchHistoryStore.setState({ watchHistory: [watchHistoryItem].concat(watchHistory) });
+    updatedHistory = [watchHistoryItem, ...watchHistory];
+    if (watchHistory.length > MAX_WATCHLIST_ITEMS_COUNT) {
+      updatedHistory = updatedHistory.slice(0, watchHistory.length - 1);
```

Comment:
Should this be `updatedHistory = updatedHistory.slice(0, MAX_WATCHLIST_ITEMS_COUNT - 1);`
Otherwise if the watchlist gets more items in it somehow it won't be restricted down to the intended length.